notes/articles/hints-principles-computer-system-design.md

```diff
@@ -454,3 +454,233 @@ Your guess about where the time is going is probably wrong. Measure before you o
+kinda like a chicken/egg problem, you want to rewrite because the code is
+not clear but you need clear code to understand what the code does so you
+can rewrite it. If you have good end to end tests or a comprehensive/correct spec/docs,
+then it maybe easier, but that is rarely the case on rewrite projects.
+
+## Software Bloat
+
+It is recurrent to find experienced software engineers to talk about software
+bloat, and it is always related to how software has much less constrains than
+most of the other engineerings:
+
+```
+Systems are complicated because it’s hard work to make them simple,
+and because people want them to do many different things.
+You can read a lot about software bloat, the proliferation of
+features in browsers and in rich applications like Word and Excel.
+
+But almost every feature has hundreds of thousands of users at least.
+The tension between keeping things simple and doing a
+lot is real, and there is no single right answer, especially for applications
+that interact with users.
+
+Still, it’s best to add features and generality slowly, because:
+
+− You’re assuming that you know the customers’ long-term needs, and you’re probably
+wrong. It’s hard enough to learn and meet their immediate needs.
+
+− It takes time to get it right, but once it’s shipped legacy
+customers make it hard to change.
+
+− More features mean more to test, and more for a bad guy to attack.
+
+So why do systems get overambitious? Because there are no clear boundaries,
+as there are with bridges for example, and programmers are creative and
+eager to tackle the next challenge. 
+```
+
+So basically software gets bloated because it can, and that doesn't seem to
+have changed in the last years, and maybe it never will.
+
+## Excessive Aspirations: Abstractions Going Bad
+
+When talking about good abstractions/components, as usual there is some
+examples of when abstractions go bad. The most classical is the extra layer
+that does almost nothing. But there are abstractions that try to do way too
+much, or do way too much at the wrong level of abstraction. Some examples:
+
+```
+The opposite of doing one thing well is doing a lot of things badly.
+Assuming reasonably competent engineering, this usually happens because
+of excessive aspirations. Some striking examples:
+
+• The original Arpanet spec mandated that a packet handed to the network
+should be delivered reliably (rather than the best-efforts delivery of
+the Internet, which is free to drop packets when the going gets tough).
+This turned out to make the network much slower than expected.
+
+• Many systems that rely on remote procedure calls have run into trouble
+because RPCs look like ordinary procedure calls, but in fact are much more
+expensive, have very unpredictable latency, and can fail completely when
+the network or the target system fails.
+```
+
+Specially on the RPC/remote objects example I sense there is some lesson to be
```

Comment:
Not sure if that makes sense, but this reminds me of the idea of formal and informal interface (from A Philosophy of Software Design). The designers of those RPC systems tried to make the formal interface as simple as possible ("it's just like a regular method call!"), but that means the informal interface became more complex. Which is not a good trade-off.

Actually, now that I think about it, I guess you're saying the same thing here: 😄 

> it is the question of "can this complexity be safely hidden ?"

Reply:
Indeed I think the points are related, feels like a leaky abstraction, the formal side of it seems great/abstracted/local, but the informal side of it leaks the networking, because there is no proper way to abstract networking latency/failures, the orders of magnitude are different on latency/reliability when compared to local method calls.

Also in this case abstracting it away makes it harder to isolate on code wher enetworking happens, the main issue is when you pass around instances to objects everywhere and it is so "well abstracted" that you are not aware which ones are local and which are not, so you may iterate on 1000 objects calling a method and now that will take forever to run, when compared to milliseconds, and there is no good way to know that by looking at code without diving into a lot of layers.

I have a similar feeling with ORMs, not necessarily bad but it is more prone to people messing around with the objects in a very non-performant way because they look just like local objects, which usually are quite fast. Languages like python can abstract this to the level of very scary, since you can override simple assignments of fields, so doing something like:

```python
cuteobj.something = "yay"
```

Maybe be a classic assigment in memory to the object...or a network call, or a disk operation...or all of the above =P. It is tough, abstractions are the base to solve complex problems..but they definitely can go wrong =P (lack of abstraction also can go wrong..but on different ways).